pkg/util/marshal/query.go

```diff
@@ -77,6 +79,14 @@ func NewStreams(s logqlmodel.Streams) (loghttp.Streams, error) {
 	ret := make([]loghttp.Stream, len(s))
 
 	for i, stream := range s {
+		// The rune error replacement is rejected by Prometheus hence replacing them with space.
+		removeInvalidUtf := func(r rune) rune {
+			if r == utf8.RuneError {
+				return 32 // rune value for space
+			}
+			return r
+		}
+		stream.Labels = string(bytes.Map(removeInvalidUtf, []byte(stream.Labels)))
```

Comment:
it's better to use `bytes.Map()` only if `[]byte(stream.Labels)` contains `utf8.RuneError` .
more details [here](https://github.com/grafana/loki/pull/14068#issuecomment-2337504002)